Added full navigation definition in UserConfiguration

adaybujeda opened the pull request:
Created PR to share discuss implementation ideas.

Fixes: https://github.com/OSC/ondemand/issues/2154

Sample configuration:

```yaml
    nav_bar:
      - title: "Help"
        links:
          - title: "Doc Links"
          - title: "how to docs"
            icon: fa://desktop
            url: "/docs"
            new_tab: false
          - title: "submit a help ticket"
            icon: fa://question-circle
            url: "/tickets"
            new_tab: false
          - app: "sys/sas"
      - title: "Documentation"
        url: "/sid2"
        new_tab: true
      - title: "Apps Menu"
        links:
          - title: "Main App"
          - app: "sys/shell"
          - title: "Sas Main"
            app: "sys/sas"
          - title: "SubApp"
            tokens: [ "sys/Rstudio/gpu" ]
          - title: "Core Apps"
          - tokens: [ "sys/Rstudio/*" ]
          - title: "Profiles"
          - title: "Team 1"
            icon: fa://question-circle
            profile: "team1"
      - title: "SAS Menu"
        app: "sys/sas"
      - title: "Wiki"
        icon: fa://desktop
        url: "/sid2"
        new_tab: true
      - title: "About"
        url: "https://www.iq.harvard.edu/research-computing#AboutSidNG"
        new_tab: false
      - template: "sessions"
      - template: "custom/nav_featured_apps"
```
Renders:

<img width="1225" alt="Screenshot 2022-09-01 at 12 10 25" src="https://user-images.githubusercontent.com/17336210/187900506-0902af0a-cdb1-4ff4-8626-eff4739e19f6.png">

<img width="639" alt="Screenshot 2022-09-01 at 12 10 33" src="https://user-images.githubusercontent.com/17336210/187900539-bdc099e4-3f9f-4299-a469-84f58324035e.png">

<img width="831" alt="Screenshot 2022-09-01 at 12 13 21" src="https://user-images.githubusercontent.com/17336210/187900881-aedfee62-8ed4-42d9-817e-c25d964ca044.png">


┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1135148780858012/1202851741436551) by [Unito](https://www.unito.io)

adaybujeda commented on apps/dashboard/app/apps/featured_app.rb:
```diff
@@ -1,21 +1,19 @@
+    @sub_feature = sub_feature.to_s
     @token = token || router.token
   end
 
```
This ^^ is an idea to fix the `categories` / `original_categories` issue when creating `FeaturedApps`, as well as making it a bit more generic to be reused.

I am using `FeaturedApps` when creating a flat list of apps and sub apps for the token matcher.

At the moment the logic is in NavBar, but it makes sense to consolidate it with the token matcher logic in `Router`

adaybujeda commented on apps/dashboard/app/apps/nav_bar.rb:
```diff
@@ -0,0 +1,109 @@
+        flat_list.concat(app.sub_app_list.map{|sub_app| FeaturedApp.new(sub_app.router, token: sub_app.token)})
+      else
+        flat_list.append(FeaturedApp.new(app.router, token: app.token))
+      end
```
Using `FeaturedApp` here to have a common interface for apps and sub apps.

Before I was returning an array with a mixture of `OodApp` and `BatchConnect::App`

adaybujeda commented on apps/dashboard/app/controllers/application_controller.rb:
```diff
@@ -57,7 +58,7 @@ def sys_app_groups
   def pinned_app_group
-    OodAppGroup.groups_for(apps: @pinned_apps, nav_limit: @user_configuration.pinned_apps_menu_length)
+    OodAppGroup.groups_for(apps: @pinned_apps, group_by: :feature, nav_limit: @user_configuration.pinned_apps_menu_length)
   end
```
This is ^^the idea to solve the `categories` / `original_categories` issue fro `FeatureApp`

adaybujeda commented on apps/dashboard/app/apps/nav_bar.rb:
```diff
@@ -0,0 +1,89 @@
+      @template ? "layouts/nav/#{@template}" : "layouts/nav/custom/#{type}"
+    end
+  end
+end
```
I have consolidated the `TokenMatcher` logic into the router class and refactored the existing `pinned_apps` methods to make more `FeaturedApp` focus

johrstrom commented:
This is a big change, so it'll take a bit for me to look it over and so on. Especially with all the other PRs in flight. Just a heads up that this may not get my full attention until next week.

adaybujeda commented on apps/dashboard/app/apps/nav_bar.rb:
```diff
@@ -0,0 +1,103 @@
+      @tokens    = config.fetch(:tokens, nil)
+      @template  =  File.join("layouts/nav", config.fetch(:template, default_template))
+
+      @title = I18n.t(@title.delete_prefix("i18n://")) if @title && @title.start_with?("i18n://")
```
This is an idea to support translations for custom titles. As mentioned in the issue, not sure how useful this might be.

johrstrom replied:
Yea I don't think this feature is wanted. You'd end up having to configure it in 2 or more places. Internationalization is about changing text you would otherwise not be able to change. But with this - you're able to put what you like directly in the YAML, so you can directly put Chinese or any language directly in the config.

I guess I'm not sure what problem it solves because the admin configuring the file can already put whatever they like here.

adaybujeda replied:
OK - Will remove

johrstrom commented on apps/dashboard/app/apps/nav_bar.rb:
```diff
@@ -0,0 +1,103 @@
+    @nav_items = nav_items
+  end
+
+  def has_items?
+    nav_items.size > 0
+  end
```
```suggestion
  def empty?
    nav_items.size == 0
  end
```

There's also a private `items?` below. Seems like all of these can coalesce to 1 API.  In any case, with the `?` we don't typically need `has` or `is` prefixes.

adaybujeda replied:
Will update

johrstrom commented on apps/dashboard/app/models/user_configuration.rb:
```diff
@@ -83,22 +89,6 @@ def navbar_type
 
-  # What to group pinned apps by
-  # @return [String, ""] Defaults to ""
-  def pinned_apps_group_by
```
We can't remove this because this is how the `pinned_apps` landing page can be grouped. You can see the group by section here in the documentation.


https://osc.github.io/ood-documentation/latest/customization.html#pinning-applications-to-the-dashboard

adaybujeda replied:
I didn't remove it, I updated the code to support 2 new properties in `FeaturedApp` (feature and sub_feature).
UserConfiguration still has `pinned_apps_group_by`:

```
# What to group pinned apps by
# @return [String, ""] Defaults to "
ConfigurationProperty.property(name: :pinned_apps_group_by, read_from_env: true, default_value: ""),
```

By adding the 2 new properties to `FeaturedApp`, there is no need to do the `original_category`, `original_subcategory` logic. When creating a FeaturedApp from a pinned_app, the code will set feature to and sub_feature to . This will have the same effect as before.

The reason for this change is simply to remove the logic for `original_category` and  `original_subcategory`.

What do you think?

johrstrom replied:
Sorry, I didn't notice that it had gotten moved. Moving it is fine, I just didn't want to _remove_ it entirely.

johrstrom commented:
OK - I took a bit of time to look this over and I've come to think that (a) we should rely on existing partials & objects when we can which leads to: (b) this may need to be broken up to iteratively coalesce existing behavior/models/templates and this new behavior.

Here's what I came up with as an example that reuses `OodAppGroup`, `OodAppLink` and the partial views for the same.

[variable_navbar_diff.txt](https://github.com/OSC/ondemand/files/9478801/variable_navbar.txt)

I was testing with this config.

```yaml
nav_bar:
  - "clusters"
  - "sys/bc_osc_rstudio"
``` 

Now that implementation lacks a lot of the functionality what you have, and that's why I'm thinking we need to break it up a bit more. 

Reusing some things (as you've likely found) is not going to be super easy, but if we attack it piece meal, feature by feature, it may be easier to shift things one way or another.

adaybujeda commented:
My concern with this approach is that it might restrict the features that we can implement as they need to fit the `Group/Link/Category/Subcategory` structure. Another concern is that once we start adding `category/subcategory` overrides, custom links, and other features, the implementation in the back-end will be complicated.

How do you see a more complex navigation configuration example with custom app titles, categories and subcategories?

This will help me with my concerns.

johrstrom commented:
You mean this structure?  

```yaml
      - title: "Apps Menu"
        links:
          - title: "Main App"
          - app: "sys/shell"
          - title: "Sas Main"
            app: "sys/sas"
          - title: "SubApp"
            tokens: [ "sys/Rstudio/gpu" ]
          - title: "Core Apps"
          - tokens: [ "sys/Rstudio/*" ]
          - title: "Profiles"
          - title: "Team 1"
            icon: fa://question-circle
            profile: "team1"
```

or maybe this bit directly?  I'll try to come up with something, but I feel like the decorator pattern can take the actual OodApp `sys/sas` and override the category and subcategory in situ. Honestly that's how the `FeatureApp` probably should have worked instead of a real subclass - just taking the original object and decorating it at runtime may have help.

```
          - title: "Sas Main"
            app: "sys/sas"
```

You're right to have concerns though. I'm just a little afraid that we're not reusing _anything_.

adaybujeda commented:
It would be great to see the same configuration (the big one) with your approach as it is a mixture of different types into the same menu item.

```yaml
      - title: "Apps Menu"
        links:
          - title: "Main App"
          - app: "sys/shell"
          - title: "Sas Main"
            app: "sys/sas"
          - title: "SubApp"
            tokens: [ "sys/Rstudio/gpu" ]
          - title: "Core Apps"
          - tokens: [ "sys/Rstudio/*" ]
          - title: "Profiles"
          - title: "Team 1"
            icon: fa://question-circle
            profile: "team1"
```

johrstrom commented:
Here's a quick hack for something similar - it does not fully support what you have here. (note I also changed `app` to `apps` and removed `tokens` because they're all kinda the same and you can cast to Array. I've left my config below).

[partial_app_and_link_support_diff.txt](https://github.com/OSC/ondemand/files/9480778/partial_app_and_link_support_diff.txt)

There are a couple of features I'm now noticing you have that I wonder about
* changing the title of the app just in the navbar. I'm not sure if we want that because there'll then be 2 names you can reference it as. Here, you're trying to rename this app `sys/sas` but it'll still show up everywhere else as the title. Could get confusing when you're trying to reference something by 2 names.
```yaml
          - title: "Sas Main"
            app: "sys/sas"
```
* The `template` support I'm unsure about. Clearly folks will need stuff like `my_interactive_sessions` but I'm wondering if just keywords are better than folks having to remember (or us document) partial paths. I'm not sure how to solve it and maybe `template` will be the best for now and we'll change it later. 

I'm not against _extending_ what we have. As an example - OodAppGroup orders automatically which isn't something we'd want here.

<details>
<summary>ondemand.d/ondemand.yml</summary>

```yaml
nav_bar:
  - title: "Help"
    links:
      - title: "Doc Links"
      - title: "how to docs"
        icon_uri: fa://desktop
        url: "/docs"
        new_tab: false
      - title: "submit a help ticket"
        icon_uri: fa://question-circle
        url: "/tickets"
        new_tab: false
      - apps: "sys/sas"
  - title: "Documentation"
    url: "/sid2"
    new_tab: true
  - title: "Apps Menu"
    links:
      - title: "Main App"
      - apps: "sys/shell"
      - title: "Sas Main"
        apps: "sys/bc_osc_rstudio_server"
      - title: "SubApp"
        apps: [ "sys/bc_desktop" ]
      - title: "Core Apps"
      - apps: [ "sys/bc_desktop" ]
      - title: "Profiles"
      - title: "Team 1"
        icon_uri: fa://question-circle
        profile: "team1"
  - title: "SAS Menu"
    apps: "sys/bc_osc_jupyter"
  - title: "Wiki"
    icon_uri: fa://desktop
    url: "/sid2"
    new_tab: true
  - title: "About"
    url: "https://www.iq.harvard.edu/research-computing#AboutSidNG"
    new_tab: false
  - template: "sessions"
  - template: "custom/nav_featured_apps"
```

</details>

johrstrom commented:
On reflection, here are some other notes.

I do want to promote reuse of existing classes/partials/concepts/etc. I'm happy to extend whatever, but I think really app groups, categories, subcategories are good concepts that we have functionality around.

As you've indicated in this PR - the concept of `FeaturedApp` has lost it's meaning a bit, or at least we're stretching it beyond what it was. That concept is something more like an app query service with the ability to overwrite the category and subcategory, we just didn't say that directly. 

So I'm hesitant to continue that mismatch with this notion of _feature_, because really it's just just _app recategorization_.

adaybujeda commented:
I added your changes into this PR to review and compare.
I have simply renamed to `NavItems` so that they can work side by side.

I have added some minor changes to your approach:
- Added support for apps as top level links.
- Fixed styling for top level links
- Added optional sorting for groups

adaybujeda commented:
Created a new `Navigation` class simply to refactor for the code for the new `NavBar` approach.
Not sure it reads better then the original, but I fixed a couple of bugs.

Found a problem with `title`. It defines the subcategory as well as the title for a link, so when link with title is defined, it creates a new section in the navigation menu.

adaybujeda commented:
I have made the necessary changes to achieve feature parity between the 2 approaches.

I am going to create a new PR with only the changes based on Groups and Links to better compare with this approach

adaybujeda commented:
I have separated the 2 implementations.

This is the new PR: https://github.com/OSC/ondemand/pull/2270 have the Group/link based implemenation.

adaybujeda commented:
I have deleted some comments, they were updates on the changes I was making to your prototype.
This is a summary:
- I have extended the prototype to match the features of my PR.
- There are certain overrides that I didn't not implement based on your comments (like app name override for title)
- I had to update `OodAppGroup` to make sorting optional.
- I updated `OodAppLink` and `_link.html.erb` to fix styling and support all properties, this requires more work.
- I created a new PR with all these changes: https://github.com/OSC/ondemand/pull/2270

This is the configuration I use for testing:

```yaml
    nav_bar:
      - title: "Help"
        links:
          - group: "Doc Links"
          - apps: "sys/sas"
          - group: "Other Links"
          - apps: "sys/sas"
          - title: "how to docs"
            icon_uri: fa://desktop
            url: "/docs"
            new_tab: false
          - title: "Submit a help ticket"
            icon_uri: fa://question-circle
            url: "/tickets"
            new_tab: false
          - apps: "sys/sas"
      - title: "Documentation"
        url: "/sid2"
        new_tab: true
      - title: "Apps Menu"
        links:
          - apps: [ "sys/Rstudio/gpu" ]
          - apps: [ "sys/sas" ]
          - title: "Sas Main"
            apps: "sys/sas"
          - group: "SubApp"
            apps: [ "sys/bc_desktop" ]
          - group: "Core Apps"
          - apps: [ "sys/bc_desktop" ]
          - title: "RStudio"
          - apps: [ "sys/Rstudio/*" ]
          - group: "Profiles"
          - title: "Team 1"
            profile: "team1"
          - title: "Team 2"
            profile: "team2"
      - title: "SAS Menu"
        apps: "sys/sas"
      - title: "Wiki"
        icon_uri: fa://desktop
        url: "/sid2"
        new_tab: true
      - title: "About"
        url: "https://www.iq.harvard.edu/research-computing#AboutSidNG"
        new_tab: false
      - template: all_apps
      - template: sessions
```

<img width="986" alt="Screenshot 2022-09-06 at 10 32 49" src="https://user-images.githubusercontent.com/17336210/188601072-fc11aed7-97cd-4178-af58-2e867d6903f8.png">

adaybujeda commented:
If you are happy with the changes I made into the new PR and using the existing `Groups` and `Links` is the general direction that you want to follow, I can discard this PR and continue the work on the new one.

adaybujeda commented:
Closing - Moving conversation to new PR: : https://github.com/OSC/ondemand/pull/2270
